README
Net-IP-XS - IPv4/IPv6 address library

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

GMP (http://gmplib.org) must be installed for this to build properly.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::IP::XS

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-IP-XS

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Net-IP-XS

    CPAN Ratings
        http://cpanratings.perl.org/d/Net-IP-XS

    Search CPAN
        http://search.cpan.org/dist/Net-IP-XS


COPYRIGHT AND LICENCE

Copyright (C) 2010 Tom Harrison <tomhrr@cpan.org>.

Original inet_pton4 and inet_pton6 functions are copyright (C) 2006
Free Software Foundation.

Original interface, and the auth and ip_auth functions, are copyright
(C) 1999-2002 RIPE NCC.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA.


Changes
Revision history for Net-IP-XS

0.08    12 May 2012
        - Remove GMP dependency.
        - Fix NI_ip_aggregate memory error on invalid input.
        - Change NI_ip_binadd's behaviour on invalid input (non-zeroes
          treated as ones).
        - NI_ip_bintoint will now return at most ((1 << 128) - 1).
        - NI_ip_inttobin's result will now always be either 32 or 128
          characters in length.

0.07    18 August 2011
        - NI_ip_compress_v4_prefix now matches the behaviour of the
          version from Net::IP (it was returning incorrect results
          when the prefix length was not on an octet boundary).

0.06    11 March 2011
        - Fixed 0.05 change (it led to a memory leak in NI_aggregate).
        - Now compiles when using a C++ compiler (tested with gcc-g++).
        - Miscellaneous problems (wrong types, unused variables).

0.05    08 March 2011
        - Missing 'return 0' statement on aggregation failure in the 
          IPv6 aggregate object method.

0.04    07 March 2011
        - Use Devel::CheckLib when installing to see whether GMP 
          is present.

0.03    26 February 2011
        - Modified NI_ip_compress_address so that it does not use '::' 
          to shorten a single 16-bit 0 field (see RFC 5952).

0.02    10 April 2010
        - Added ip_add_num function and +/bool overloads.
        - Added 'PROC' export tag.

0.01    06 April 2010
        - Initial release.


object.h
/*
object.h - Functions for Net::IP::XS's object-oriented interface.

Copyright (C) 2010 Tom Harrison <tomhrr@cpan.org>
Original inet_pton4, inet_pton6 are Copyright (C) 2006 Free Software
Foundation.
Original interface, and the auth and ip_auth functions, are Copyright
(C) 1999-2002 RIPE NCC.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#ifndef NETIP_OBJECT
#define NETIP_OBJECT

#ifdef __cplusplus
extern "C" {
#endif

#include "gmp.h"
#include "limits.h"

void NI_object_set_Error_Errno(SV *ipo, int Errno, char *Error, ...);
void NI_copy_Error_Errno(SV *ipo);

int  NI_find_prefixes(SV *ipo, char **prefixes, int *pcount);

int  NI_set_ipv6_mpzs(SV *ipo);
int  NI_set(SV* ip, char *data, int ipversion);

int  NI_get_begin_mpz(SV *ipo, mpz_t **begin);
int  NI_get_end_mpz(SV *ipo, mpz_t **end);
int  NI_get_mpzs(SV *ipo, mpz_t **begin, mpz_t **end);

int  NI_short(SV *ipo, char *buf);
int  NI_print(SV *ipo, char *buf, int maxlen);

int  NI_last_ip(SV *ipo, char *buf, int maxlen);

int  NI_size_str_ipv4(SV *ipo, char *buf);
int  NI_size_str_ipv6(SV *ipo, char *buf);
int  NI_size_str(SV *ipo, char *size);

int  NI_intip_str_ipv4(SV *ipo, char *buf);
int  NI_intip_str_ipv6(SV *ipo, char *buf);
int  NI_intip_str(SV *ipo, char *buf, int maxlen);

int  NI_hexip_ipv4(SV *ipo, char *buf);
int  NI_hexip_ipv6(SV *ipo, char *hexip);
int  NI_hexip(SV *ipo, char *buf, int maxlen);

int  NI_hexmask(SV *ipo, char *buf, int maxlen);

int  NI_prefix(SV *ipo, char *buf, int maxlen);

int  NI_mask(SV *ipo, char *buf, int maxlen);

int  NI_iptype(SV *ipo, char *buf, int maxlen);

int  NI_reverse_ip(SV *ipo, char *buf);

int  NI_last_bin(SV *ipo, char *buf, int maxlen);

int  NI_last_int_str_ipv4(SV *ipo, char *buf);
int  NI_last_int_str_ipv6(SV *ipo, char *buf);
int  NI_last_int_str(SV *ipo, char *buf, int maxlen);

int  NI_bincomp(SV *ipo1, char *op, SV *ipo2, int *resbuf);

int  NI_overlaps_ipv4(SV *ipo1, SV* ipo2, int *buf);
int  NI_overlaps_ipv6(SV *ipo1, SV* ipo2, int *buf);
int  NI_overlaps(SV *ipo1, SV* ipo2, int *buf);

SV *NI_binadd(SV *ipo1, SV *ipo2);

SV *NI_aggregate_ipv4(SV *ipo1, SV *ipo2);
SV *NI_aggregate_ipv6(SV *ipo1, SV *ipo2);
SV *NI_aggregate(SV *ipo1, SV *ipo2);

int NI_ip_add_num_ipv4(SV *ipo, unsigned long num, char *buf);
int NI_ip_add_num_ipv6(SV *ipo, mpz_t num, char *buf);
SV *NI_ip_add_num(SV *ipo, const char *num);

#ifdef __cplusplus
}
#endif

#endif


functions.h
/*
functions.h - Core functions for Net::IP::XS.

Copyright (C) 2010 Tom Harrison <tomhrr@cpan.org>
Original inet_pton4, inet_pton6 are Copyright (C) 2006 Free Software
Foundation.
Original interface, and the auth and ip_auth functions, are Copyright
(C) 1999-2002 RIPE NCC.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

#ifndef NETIP_FUNCTIONS
#define NETIP_FUNCTIONS

#include "gmp.h"
#include "limits.h"

#ifdef __cplusplus
extern "C" {
#endif

#define STRTOL_FAILED(ret, str, end) \
    (((((ret) == LONG_MAX || ((ret) == LONG_MIN)) && (errno == ERANGE)) \
        || (((ret) == 0) && ((str) == (end)))))

#define STRTOUL_FAILED(ret, str, end) \
    (((((ret) == ULONG_MAX || ((ret) == 0)) && (errno == ERANGE)) \
        || (((ret) == 0) && ((str) == (end)))))

/* Comparator constants (less than, less than or equal to, etc.). */
#define LT 1
#define LE 2
#define GT 3
#define GE 4

/* Overlap constants. */
#define IP_NO_OVERLAP       0
#define IP_PARTIAL_OVERLAP  1
#define IP_A_IN_B_OVERLAP  -1
#define IP_B_IN_A_OVERLAP  -2
#define IP_IDENTICAL       -3

/* String length constants. */
#define MAX_IPV4_STR_LEN       16
#define MAX_IPV4_RANGE_STR_LEN 19
#define IPV4_BITSTR_LEN        33
#define MAX_IPV4_REVERSE_LEN   30 
#define MAX_IPV6_STR_LEN       64
#define MAX_IPV6_RANGE_STR_LEN 68
#define IPV6_BITSTR_LEN        129
#define MAX_IPV6_REVERSE_LEN   74
#define MAX_IPV6_NORMAL_RANGE  82
#define MAX_IPV6_HEXIP_STR_LEN 35

#define MAX_TYPE_STR_LEN       256
#define MAX_PREFIXES           128
#define MAX_IPV6_NUM_STR_LEN   40

const char    *NI_hv_get_pv(SV *object, const char *key, int keylen);
int            NI_hv_get_iv(SV *object, const char *key, int keylen);
unsigned int   NI_hv_get_uv(SV *object, const char *key, int keylen);

void           NI_set_Errno(int Errno);
int            NI_get_Errno(void);
void           NI_set_Error(const char *Error);
const char    *NI_get_Error(void);
void           NI_set_Error_Errno(int Errno, const char *Error, ...);

int            NI_hdtoi(char c);
int            NI_trailing_zeroes(unsigned long n);
unsigned long  NI_bintoint(const char *bitstr, int len);
unsigned long  NI_ip_uchars_to_ulong(unsigned char uchars[4]);
void           NI_ip_uchars_to_mpz(unsigned char uchars[16], mpz_t *num);

int            NI_iplengths(int version);

void           NI_ip_mpztobin(mpz_t num, int len, char *buf);
void           NI_ip_bintompz(const char *bitstr, int len, mpz_t *num);

int            NI_ip_inttobin_str(const char *ip_int_str, 
                                  int version, char *buf);
int            NI_ip_bintoint_str(const char *bitstr, char *buf);

int            NI_ip_is_ipv4(const char *str);
int            NI_ip_is_ipv6(const char *str);
int            NI_ip_get_version(const char *str);
int            NI_ip_get_mask(int len, int version, char *buf);

int            NI_ip_last_address_ipv6(mpz_t ip, int len, mpz_t *buf);
unsigned long  NI_ip_last_address_ipv4(unsigned long ip, int len);
int            NI_ip_last_address_bin(const char *bitstr, int len, 
                                      int version, char *buf);

int            NI_ip_bincomp(const char *bitstr_1, const char *op_str, 
                             const char *bitstr_2, int *result);

void           NI_ip_is_overlap_ipv6(mpz_t begin_1, mpz_t end_1,
                                     mpz_t begin_2, mpz_t end_2, int *result);
void           NI_ip_is_overlap_ipv4(unsigned long begin_1, 
                                     unsigned long end_1,
                                     unsigned long begin_2, 
                                     unsigned long end_2,
                                     int *result);
int            NI_ip_is_overlap(const char *begin_1, const char *end_1,
                                const char *begin_2, const char *end_2, 
                                int *result);

int            NI_ip_check_prefix_ipv6(mpz_t ip, int len);
int            NI_ip_check_prefix_ipv4(unsigned long ip, int len);
int            NI_ip_check_prefix(const char *bitstr, int len, int version);

void           NI_ip_get_prefix_length_ipv6(mpz_t mpz_1, mpz_t mpz_2, 
                                            int bits, int *len);
void           NI_ip_get_prefix_length_ipv4(unsigned long begin, 
                                            unsigned long end,
                                            int bits, int *len);
int            NI_ip_get_prefix_length(const char *bitstr_1, 
                                       const char *bitstr_2, 
                                       int *len);

void           NI_ip_inttoip_ipv6(unsigned long n1, unsigned long n2, 
                                  unsigned long n3, unsigned long n4, 
                                  char *buf);
void           NI_ip_inttoip_ipv4(unsigned long n, char *buf);
void           NI_ip_inttoip_mpz(mpz_t ip, char *buf);
int            NI_ip_bintoip(const char *bitstr, int version, char *buf);

int            NI_ip_binadd(const char *first, const char *second, 
                            char *buf, int maxlen);

int            NI_ip_range_to_prefix_ipv6(mpz_t begin, mpz_t end,
                                          int version, char **prefixes, 
                                          int *pcount);
int            NI_ip_range_to_prefix_ipv4(unsigned long begin, 
                                          unsigned long end,
                                          int version, char **prefixes, 
                                          int *pcount);
int            NI_ip_range_to_prefix(const char *bitstr_1, 
                                     const char *bitstr_2,
                                     int version, char **prefixes, 
                                     int *pcount);

int            NI_ip_aggregate_tail(int res, char **prefixes, int pcount,
                                    int version, char *buf);
int            NI_ip_aggregate_ipv6(mpz_t b1, mpz_t e1, mpz_t b2, mpz_t e2,
                                    int ipversion, char *buf);
int            NI_ip_aggregate_ipv4(unsigned long b1, unsigned long e1,
                                    unsigned long b2, unsigned long e2,
                                    int ipversion, char *buf);
int            NI_ip_aggregate(const char *b1, const char *e1, 
                               const char *b2, const char *e2,
                               int ipversion, char *buf);

int            NI_ip_iptobin(const char *ip, int ipversion, char *buf);

int            NI_ip_expand_address_ipv6(const char *ip, char *retbuf);
int            NI_ip_expand_address_ipv4(const char *ip, char *buf);
int            NI_ip_expand_address(const char *ip, int version, char *buf);

int            NI_ip_reverse_ipv6(const char *ip, int len, char *buf);
int            NI_ip_reverse_ipv4(const char *ip, int len, char *buf);
int            NI_ip_reverse(const char *ip, int len, int ipversion, char *buf);

int            NI_ip_normalize_prefix_ipv6(mpz_t ip, char *slash,
                                           char *ip1buf, char *ip2buf);
int            NI_ip_normalize_prefix_ipv4(unsigned long ip, char *slash,
                                           char *ip1buf, char *ip2buf);
int            NI_ip_normalize_prefix(char *ip, char *ip1buf, char *ip2buf);

int            NI_ip_tokenize_on_char(char *str, char separator,
                                      char **end_first, char **second);
int            NI_ip_normalize_plus_ipv6(char *ip, char *num,
                                         char *ipbuf1, char *ipbuf2);
int            NI_ip_normalize_plus_ipv4(char *ip, char *num,
                                         char *ipbuf1, char *ipbuf2);
int            NI_ip_normalize_range(char *ip, char *ipbuf1, char *ipbuf2);

int            NI_ip_normalize_plus(char *ip1, char *ipbuf1, char *ipbuf2);

int            NI_ip_normalize_bare(char *ip, char *ipbuf1);

int            NI_ip_normalize(char *ip, char *ipbuf1, char *ipbuf2);

int            NI_ip_normal_range(char *ip, char *buf);

int            NI_ip_compress_v4_prefix(const char *ip, int len, 
                                        char *buf, int maxlen);
int            NI_ip_compress_address(const char *ip, int version, char *buf);

int            NI_ip_splitprefix(const char *prefix, char *ipbuf, int *lenbuf);
int            NI_ip_iptype(const char *ip, int version, char *buf);
int            NI_ip_is_valid_mask(const char *mask, int version);
int            NI_ip_prefix_to_range(const char *ip, int len, int version, 
                                     char *buf);
int            NI_ip_get_embedded_ipv4(const char *ipv6, char *buf);

#ifdef __cplusplus
}
#endif

#endif


lib/Net/IP/XS.pm
package Net::IP::XS;

use warnings;
use strict;

use 5.006;

use Math::BigInt;
use Tie::Hash::Sorted;
use Tie::Simple;

our $VERSION = '0.08';

our $IP_NO_OVERLAP      = 0;
our $IP_PARTIAL_OVERLAP = 1;
our $IP_A_IN_B_OVERLAP  = -1;
our $IP_B_IN_A_OVERLAP  = -2;
our $IP_IDENTICAL       = -3;

our @EXPORT_OK = qw(Error
                    Errno
                    ip_iptobin
                    ip_bintoip
                    ip_iplengths
                    ip_bintoint
                    ip_inttobin
                    ip_expand_address
                    ip_is_ipv4
                    ip_is_ipv6
                    ip_get_version
                    ip_get_mask
                    ip_last_address_bin
                    ip_splitprefix
                    ip_is_valid_mask
                    ip_bincomp
                    ip_binadd
                    ip_get_prefix_length
                    ip_compress_v4_prefix
                    ip_is_overlap
                    ip_check_prefix
                    ip_range_to_prefix
                    ip_get_embedded_ipv4
                    ip_aggregate
                    ip_prefix_to_range
                    ip_reverse
                    ip_normalize
                    ip_compress_address
                    ip_iptype
                    ip_auth
                    ip_normal_range
                    $IP_NO_OVERLAP
                    $IP_PARTIAL_OVERLAP
                    $IP_A_IN_B_OVERLAP
                    $IP_B_IN_A_OVERLAP
                    $IP_IDENTICAL);

our %EXPORT_TAGS = (PROC => [@EXPORT_OK]);

use overload (
    '+'    => 'ip_add_num',
    'bool' => sub { @_ },
);

use base qw(DynaLoader Exporter);
__PACKAGE__->bootstrap($VERSION);

our $ERROR;
our $ERRNO;

BEGIN {
    tie $ERROR, 'Tie::Simple', 1, FETCH => \&ip_get_Error,
                                  STORE => \&ip_set_Error;
    tie $ERRNO, 'Tie::Simple', 1, FETCH => \&ip_get_Errno,
                                  STORE => \&ip_set_Errno;
};

our %IPv4ranges = (
    '00000000'                 => 'PRIVATE',     # 0/8
    '00001010'                 => 'PRIVATE',     # 10/8
    '01111111'                 => 'PRIVATE',     # 127.0/8
    '101011000001'             => 'PRIVATE',     # 172.16/12
    '1100000010101000'         => 'PRIVATE',     # 192.168/16
    '1010100111111110'         => 'RESERVED',    # 169.254/16
    '110000000000000000000010' => 'RESERVED',    # 192.0.2/24
    '1110'                     => 'RESERVED',    # 224/4
    '11110'                    => 'RESERVED',    # 240/5
    '11111'                    => 'RESERVED',    # 248/5
);

our %IPv6ranges = (
    '00000000'   => 'RESERVED',                  # ::/8
    '00000001'   => 'RESERVED',                  # 0100::/8
    '0000001'    => 'RESERVED',                  # 0200::/7
    '000001'     => 'RESERVED',                  # 0400::/6
    '00001'      => 'RESERVED',                  # 0800::/5
    '0001'       => 'RESERVED',                  # 1000::/4
    '001'        => 'GLOBAL-UNICAST',            # 2000::/3
    '010'        => 'RESERVED',                  # 4000::/3
    '011'        => 'RESERVED',                  # 6000::/3
    '100'        => 'RESERVED',                  # 8000::/3
    '101'        => 'RESERVED',                  # A000::/3
    '110'        => 'RESERVED',                  # C000::/3
    '1110'       => 'RESERVED',                  # E000::/4
    '11110'      => 'RESERVED',                  # F000::/5
    '111110'     => 'RESERVED',                  # F800::/6
    '1111101'    => 'RESERVED',                  # FA00::/7
    '1111110'    => 'UNIQUE-LOCAL-UNICAST',      # FC00::/7
    '111111100'  => 'RESERVED',                  # FE00::/9
    '1111111010' => 'LINK-LOCAL-UNICAST',        # FE80::/10
    '1111111011' => 'RESERVED',                  # FEC0::/10
    '11111111'   => 'MULTICAST',                 # FF00::/8
    '00100000000000010000110110111000' => 'RESERVED',    # 2001:DB8::/32

    '0' x 96 => 'IPV4COMP',                              # ::/96
    ('0' x 80) . ('1' x 16) => 'IPV4MAP',                # ::FFFF:0:0/96

    '0' x 128         => 'UNSPECIFIED',                  # ::/128
    ('0' x 127) . '1' => 'LOOPBACK'                      # ::1/128
);

sub sort_by_length
{
    my ($hash) = @_;

    [ sort { length($b) <=> length($a) } keys %{$hash} ];
}

BEGIN {
    tie %IPv4ranges, 'Tie::Hash::Sorted',
        'Hash'         => \%IPv4ranges,
        'Sort_Routine' => \&sort_by_length,
        'Optimization' => 'keys';

    tie %IPv6ranges, 'Tie::Hash::Sorted',
        'Hash'         => \%IPv6ranges,
        'Sort_Routine' => \&sort_by_length,
        'Optimization' => 'keys';
};

sub Error       { $ERROR }
sub Errno       { $ERRNO }

sub ip_bintoint { Math::BigInt->new(ip_bintoint_str($_[0]))        }
sub ip_inttobin { ip_inttobin_str(Math::BigInt->new($_[0]), $_[1]) }

sub binip       { $_[0]->{'binip'}     }
sub last_bin    { $_[0]->{'last_bin'}  }
sub version     { $_[0]->{'ipversion'} }
sub error       { $_[0]->{'error'}     }
sub errno       { $_[0]->{'errno'}     }
sub prefixlen   { $_[0]->{'prefixlen'} }
sub ip          { $_[0]->{'ip'}        }
sub is_prefix   { $_[0]->{'is_prefix'} }
sub binmask     { $_[0]->{'binmask'}   }

sub size        { Math::BigInt->new(size_str($_[0]))     }
sub intip       { Math::BigInt->new(intip_str($_[0]))    }
sub last_int    { Math::BigInt->new(last_int_str($_[0])) }

sub auth 
{
    my ($self) = shift;

    return ($self->{auth}) if defined($self->{auth});

    my $auth = ip_auth($self->ip, $self->version);

    if (!$auth) {
        $self->{error} = $ERROR;
        $self->{errno} = $ERRNO;
        return;
    }

    $self->{auth} = $auth;

    return ($self->{auth});
}

sub ip_auth 
{
    my ($ip, $ip_version) = (@_);

    if (not $ip_version) {
        $ERROR = "Cannot determine IP version for $ip";
        $ERRNO = 101;
        return;
    }

    if ($ip_version != 4) {
        $ERROR = "Cannot get auth information: Not an IPv4 address";
        $ERRNO = 308;
        return;
    }

    require IP::Authority;

    my $reg = new IP::Authority;

    return ($reg->inet_atoauth($ip));
}

sub STORABLE_freeze
{
    my ($self, $cloning) = @_;

    my %copy = %{$self};
    delete @copy{qw(xs_v6_ip0 xs_v6_ip1)};

    return (1, \%copy);
}

sub STORABLE_thaw
{
    my ($self, $cloning, $dummy, $copy) = @_;

    my @fields = keys %{$copy};
    @{$self}{@fields} = @{$copy}{@fields};

    if ($self->{'ipversion'} == 6) {
        my $res = $self->set_ipv6_n128s();
        if (!$res) {
            die "Unable to generate internal IPv6 bignums.";
        }
    }
    
    return $copy;
}

1;

__END__

=head1 NAME

Net::IP::XS - XS implementation of Net::IP

=head1 SYNOPSIS

  use Net::IP::XS;

  my $ip = new Net::IP::XS ('193.0.1/24') or die (Net::IP::XS::Error());
  print ("IP  : ".$ip->ip()."\n");
  print ("Sho : ".$ip->short()."\n");
  print ("Bin : ".$ip->binip()."\n");
  print ("Int : ".$ip->intip()."\n");
  print ("Mask: ".$ip->mask()."\n");
  print ("Last: ".$ip->last_ip()."\n");
  print ("Len : ".$ip->prefixlen()."\n");
  print ("Size: ".$ip->size()."\n");
  print ("Type: ".$ip->iptype()."\n");
  print ("Rev:  ".$ip->reverse_ip()."\n");

=head1 DESCRIPTION

An XS implementation of L<Net::IP|Net::IP>. See L<Net::IP|Net::IP>'s
documentation (as at version 1.25) for the functions and methods that
are available.

=head1 DIFFERENCES BETWEEN NET::IP AND NET::IP::XS

=over 4

=item Exports

Nothing is exported by default.

=item Error messages

In some instances this won't set error codes or messages where
C<Net::IP> would, though it should be mostly the same.

=item Object-oriented interface

The object-oriented interface uses function calls and hashref lookups
internally, such that subclassing C<Net::IP::XS> will not have the
same effect as it does with C<Net::IP>.

=item ip_auth

Returns C<undef> on failure, instead of dying.

=item ip_binadd

Returns C<undef> if either of the bitstring arguments is more than 128
characters in length.

Any character of the bitstring that is not a 0 is treated as a 1. The
C<Net::IP> version returns different results for different digits, and
treats non-digits as 0.

=item ip_bintoint

The integer returned will be at most ((1 << 128) - 1) (i.e. the
largest possible IPv6 address). C<Net::IP> handles bitstrings of
arbitrary length.

=item ip_compress_address

Returns C<undef> if the IPv6 address argument is invalid.

=item ip_compress_v4_prefix

Returns C<undef> if the C<len> argument is negative or greater than
32.

=item ip_expand_address

Does not set C<Error> or C<Errno> where there is a problem with an
embedded IPv4 address within an IPv6 address. 

Returns the zero IP address if the empty string is provided. The
C<Net::IP> version returns C<undef>. 

Returns a full IPv6 address if a partial address is provided (e.g.
returns 'ffff:ffff:0000:0000:0000:0000:0000:0000' if 'ffff:ffff' is
provided).  The C<Net::IP> version returns the partial address. 

Returns C<undef> on an invalid IPv4/IPv6 address. The C<Net::IP>
version returns the zero address for IPv4 and whatever was provided
for IPv6.

=item ip_get_mask

The mask returned will always have a length equal to the number of
bits in an address of the specified IP version (e.g. an IPv4 mask will
always comprise 32 characters). The C<Net::IP> version will return a
longer mask when the C<len> argument is larger than the number of bits
in the specified IP version. 

If a negative C<len> is provided, it will be treated as zero.

=item ip_inttobin

The bitstring returned will always be either 32 or 128 characters in
length, and it returns C<undef> if the integer argument would require
more than 128 characters to represent as a bitstring. If an invalid
version is provided, the returned bitstring will be 128 characters in
length. The C<Net::IP> version handles arbitrary integers and expands
to accommodate those integers, regardless of the version argument.
Also, if an invalid version is provided, the returned bitstring is
only as long as is necessary to accommodate the bitstring.

=item ip_iptobin

Returns C<undef> on an invalid IPv4/IPv6 address.

=item ip_last_address_bin

Returns an empty string if an invalid version (i.e. not 4 or 6) is
provided. If the bitstring provided is longer than the number of bits
in the specified version, then only the first 32/128 bits will be used
in determining the last address. If the C<len> provided is invalid
(negative or more than 32/128 depending on the version), it will be
treated as the maximum length of the specified version.

=item ip_normalize

For the 'plus' style of string (e.g. '1.0.0.0 + 255'), whitespace
between the plus character and the parts before and after it is
optional. In the C<Net::IP> version, there has to be some whitespace
before and after the plus character. Also, C<undef> will be returned
if the part after the plus sign is not a number. The C<Net::IP> version
will return two copies of the single address in this instance.

For the 'prefix range' style of string (e.g. '1.0.0.0/8'), the part
after the slash must be a number. If it is not, C<undef> will be
returned. The C<Net::IP> version will return two copies of the single
address in this instance.

=item ip_range_to_prefix

Returns C<undef> if the version argument is invalid.

=item ip_reverse

The C<len> argument determines the length of the reverse domain -
e.g., if the arguments are '127.0.0.1', '16' and '4', the reverse
domain will be '0.127.in-addr.arpa.'. The C<Net::IP> version does not
take the C<len> argument into account for IPv4 addresses. For IPv6
addresses, a compressed IP address string may be provided.

=item ip_splitprefix

Returns C<undef> unless the first component of the string is less than
or equal to 64 characters in length. The C<Net::IP> version handles
strings of arbitrary length.

=item prefix

Returns a string with a prefix length of zero (e.g. '127.0.0.1/0')
where C<prefixlen> is not defined in the object. The C<Net::IP>
version will not include any prefix length in the returned string
(e.g. '127.0.0.1/').

=back

=head1 AUTHOR

Tom Harrison, C<< <tomhrr@cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-net-ip-xs at rt.cpan.org>, 
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-IP-XS>.

=head1 ACKNOWLEDGEMENTS

Manuel Valente (C<< <manuel.valente@gmail.com> >>) and the other
authors of L<Net::IP|Net::IP>.

=head1 SEE ALSO

L<Net::IP|Net::IP>, L<IP::Authority|IP::Authority>.

=head1 COPYRIGHT & LICENSE

Copyright (C) 2010-2012 Tom Harrison <tomhrr@cpan.org>.

Original inet_pton4 and inet_pton6 functions are copyright (C) 2006 
Free Software Foundation.

Original interface, and the auth and ip_auth functions, are copyright
(C) 1999-2002 RIPE NCC.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA.

=cut
